hardware/kubernetes.go

```diff
@@ -0,0 +1,132 @@
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	tinkv1alpha1 "github.com/tinkerbell/tink/pkg/apis/core/v1alpha1"
+	tink "github.com/tinkerbell/tink/pkg/controllers"
+	"k8s.io/client-go/rest"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ Client = &KubernetesClient{}
+
+type ListerClient interface {
+	List(ctx context.Context, list crclient.ObjectList, opts ...crclient.ListOption) error
+}
+
+// KubernetesClient is a hardware client backed by a KubernetesClient cluster that contains hardware resources.
+type KubernetesClient struct {
+	client           ListerClient
+	close            func()
+	waitForCacheSync func(context.Context) bool
+}
+
+// NewKubernetesClientOrDie creates a new KubernetesClient client. It panics upon error.
+func NewKubernetesClientOrDie(config *rest.Config, namespace string) *KubernetesClient {
+	client, err := NewKubernetesClient(config, namespace)
+	if err != nil {
+		panic(err)
+	}
+	return client
+}
+
+// NewKubernetesClient creates a new KubernetesClient client instance. It launches a goroutine to perform synchronization
+// between the cluster and internal caches. Consumers can wait for the initial sync using WaitForCachesync().
+// See k8s.io/client-go/tools/clientcmd for constructing *rest.Config objects.
+func NewKubernetesClient(config *rest.Config, namespace string) (*KubernetesClient, error) {
+	opts := tink.GetServerOptions()
+	opts.Namespace = namespace
+
+	// Use a manager from the tink project so we can take advantage of the indexes and caching it configures.
+	// Once started, we don't really need any of the manager capabilities hence we don't store it in the
+	// KubernetesClient
+	manager, err := tink.NewManager(config, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	managerCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		if err := manager.Start(managerCtx); err != nil {
+			panic(err)
+		}
+	}()
+
+	client := NewKubernetesClientWithClient(manager.GetClient())
+	client.close = cancel
+	client.waitForCacheSync = manager.GetCache().WaitForCacheSync
+
+	return client, nil
+}
+
+// NewKubernetesClientWithClient creates a new KubernetesClient instance that uses client to find resources. The
+// Close() and WaitForCacheSync() methods of the returned client are noops.
+func NewKubernetesClientWithClient(client ListerClient) *KubernetesClient {
+	return &KubernetesClient{
+		client:           client,
+		close:            func() {},
+		waitForCacheSync: func(context.Context) bool { return true },
+	}
+}
+
+// Close stops synchronization with the cluster. Subsequent calls to all of k's methods have undefined behavior.
+func (k *KubernetesClient) Close() {
+	k.close()
+}
+
+// WaitForCacheSync waits for the internal client cache to synchronize.
+func (k *KubernetesClient) WaitForCacheSync(ctx context.Context) bool {
+	return k.waitForCacheSync(ctx)
+}
+
+// IsHealthy always returns true.
+func (k *KubernetesClient) IsHealthy(context.Context) bool {
+	return true
```

Comment:
Suggestions on what to do here as its not true. Initially I thought a mutex protected `k.close` then set `k.close = nil` in the `Close()` func with a `return k.close != nil` for `IsHealthy()`. @micahhausler is there a better way in the manager to check if we're 'connected'?

Reply:
It doesn't look like controller-runtime's client exposes any health metrics.I think this is good enough